k2/csrc/fsa_utils.cu

```diff
@@ -0,0 +1,328 @@
+/**
+ * @brief Utilities for reading, writing and creating FSAs.
+ *
+ * @copyright
+ * Copyright (c)  2020  Mobvoi Inc.        (authors: Fangjun Kuang)
+ *
+ * @copyright
+ * See LICENSE for clarification regarding multiple authors
+ */
+
+#include <sstream>
+#include <utility>
+#include <vector>
+
+#include "k2/csrc/context.h"
+#include "k2/csrc/fsa_utils.h"
+
+namespace k2 {
+
+// field separator within a line for a text form FSA
+static constexpr const char *kDelim = " \t";
+
+// Convert a string to an integer. Abort the program on failure.
+static int32_t StringToInt(const std::string &s) {
+  K2_CHECK_EQ(s.empty(), false);
+
+  bool ok = false;
+  char *p = nullptr;
+  // std::strtol requires a `long` type
+  long n = std::strtol(s.c_str(), &p, 10);  // NOLINT
+  if (*p == '\0') ok = true;
+
+  auto res = static_cast<int32_t>(n);
+  if (n != res) ok = false;  // out of range
+
+  if (ok == false)
+    K2_LOG(FATAL) << "Failed to convert " << s << " to an integer";
+
+  return res;
+}
+
+// Convert a string to a float. Abort the program on failure.
+static float StringToFloat(const std::string &s) {
+  K2_CHECK_EQ(s.empty(), false);
+  char *p = nullptr;
+  float f = std::strtof(s.c_str(), &p);
+  if (*p != '\0') K2_LOG(FATAL) << "Failed to convert " << s << " to a float";
+  return f;
+}
+
+// Trim leading and trailing spaces of a string.
+static void TrimString(std::string *s) {
+  K2_CHECK_NE(s, nullptr);
+  auto not_space = [](int32_t c) { return std::isspace(c) == 0; };
+
+  s->erase(s->begin(), std::find_if(s->begin(), s->end(), not_space));
+  s->erase(std::find_if(s->rbegin(), s->rend(), not_space).base(), s->end());
+}
+
+/* Split a string to a vector of strings using a set of delimiters.
+
+   Example usage:
+
+   @code
+    std::string in = "1 2 3";
+    const char *delim = " \t";
+    std::vector<std::string> out;
+    SplitStringToVector(in, delim, &out);
+   @endcode
+
+   @param [in]  in    The input string to be split.
+   @param [in]  delim A string of delimiters.
```

Comment:
seems we only support `delim` to be a character? or the code can be run with `char sequence`?

Reply:
https://en.cppreference.com/w/cpp/string/basic_string/find_first_of

(3)

It is a set of characters.

Reply:
Thanks for the explanation.